### DISCUSSION TITLE
Suggestion: trains should check and reserve blocks instead of checking signals

### DISCUSSION THREAD

**Discussion by hairintd** [voxeltycoon/issues#745] [General]
At this time it looks for me that trains wait on pre-signals for the opening _signal_ (latest signal in the row) but not for block it opens being released. It behave more like train driven by real person who can check what he's eyes see, may be that is an idea. 
But in other hand such a behavior gives a problems on real loaded ways, see picture below.

![20210512125415_1](https://user-images.githubusercontent.com/10298859/117964872-d47ea500-b32a-11eb-9bb9-06e919038f64.jpg)

You can see junction where two rail pathes combines into one path. Each path have two railways in opposite direction. Train 45 (number sign cleared due to markup) comes from the left side and Train 1 comes from down side both moving to the Magenta block and next to the Station.
If they come to the junction at same time they need to share the Green block, which is limited by entering Signal 2 from left, 3 from down and exiting Signal 4 from top. 
Also there is Yellow block before the Green, which must be crossed by Train 45 and which is shared by trains going opposite way from signal 5 side.

Trains moves in this sequence:
1) Train 45 comes to the Signal 1, which is green and enters the Yellow block.
2) Train 1 comes to the Signal 3, which is green and enters Green block. 
3) Train 45 comes to the Signal 2, which is red due to Train 1 is in the block, and stops waiting Train 1 to leave. 
In this way we got problem when the Train 45 waits at Signal 2 it blocks way for trains moving opposite direction through the Yellow block from the Signal 5 side. The Train 45 will free Yellow block only when Magenta block will be free so this is pretty much long time for highloaded railways and potentially would lead to deadlock of whole track part.

There is simple solution for this problem. Instead of checking entering _signals_ trains should reserve and check _blocks_ next to the signal for being reserved by other trains. In case of pre-signal trains should check for whole blocks chain from entering to exiting signals. When train leaves reserved block it clear reservation so other trains can reserve this block.
Lets see how it could work in our case:
1) Train 45 comes to the Signal 1, checks Yellow block, then Green and Magenta blocks for not being locked by other trains. It's true, so train reserves all of them and keep moving. 
2) Train 1 comes to the Signal 3, checks whether Green block is free. It's reserved by the train 45 (even if it's in the Yellow block at this time), so it stops and wait until the Green and Magenta blocks will be free. 

If Train 1 comes to the Signal 3 first the sequence will be:
1) Train 1 comes to the Signal 3 checks Green then Magenta blocks. They are free so it reserves them and keep moving.
2) Train 45 comes to the Signal 1 and checks if Yellow block is free. Then it checks whether Green block is free (which is not) and stops at the Signal 1, waiting for all the blocks Yellow, Green and Magenta will be free. 

You can check implementation of this solution in other games such as factorio, transport fever, Open TTD (AFAIR), W&R:Soviet republic. It works pretty well.

**Comment by Road-Train** [comment]
Okay, I decided to take your picture and turn it into a majestic paint example, shown below:
![signals](https://user-images.githubusercontent.com/43598985/118111829-875f0980-b3e4-11eb-8461-e8ffa4535b96.png)
The lines here are your blocks, the p's are pre signals, the s's are your normal signals, their colour corresponds to the block they're checking.
The issue you described is:
Train A approaches the junction from the purple(bottom left) line at the same time as train B coming from the blue(bottom right) line. They're both trying to get to the grey/magenta esque(top right) line.

Train A arrives at it's yellow pre signal and is cleared to enter the block. Train B arrives at it's light green pre signal before train A, and is allowed through. Train A now blocks the yellow block. 

The current ingame way of preventing this from happening would be to delete the yellow pre signal that connects the light blue line to the yellow block, and the light green pre signal connecting the yellow block to the light green block, but this causes an inefficiency.

Let's say a new train A is approaching from the top left trying to go to the bottom left. A train B is coming from the bottom left trying to go top right. These trains will block one another as the entire junction is now one big block.

You're suggesting a reservation system which would have the following effect
Train A approaches the junction from the purple(bottom left) line at the same time as train B coming from the blue(bottom right) line. They're both trying to get to the grey/magenta esque(top right) line.

Train A arrives at it's yellow pre signal and reserves the yellow and the light green block. Train B could enter the light green block, but it has been reserved. The train therefore waits.

I suggest giving the pre-signal and the normal signal an extra light to reflect a 'reserved' status. In another game which has this implemented, the reserved status is yellow, and the pre signal has a fourth blue light instead.

Pre signal would be structured as follows:
red
blue
yellow
green
Signals would inherit the skin of the pre signals. This'd be confusing at first but with time people will get used to it. 
I explain my suggestion more here: https://github.com/voxeltycoon/issues/discussions/750

**Comment by hungrycat** [comment]
Note this also leads to dead-locks, even if pre-signals and signals are used to actually prevent dead-locks, e.g. train leaving a shared platform with another incoming train just having decided to use this platform. As such, I would even consider it a missing feature bordering with a bug:) There is a related discussion here: [527](https://github.com/voxeltycoon/issues/discussions/527)
